src/Microsoft.AspNet.Mvc.Core/ModelBinding/CollectionModelBinder.cs

```diff
@@ -31,16 +31,19 @@ public class CollectionModelBinder<TElement> : IModelBinder
             var valueProviderResult = await bindingContext.ValueProvider.GetValueAsync(bindingContext.ModelName);
 
             IEnumerable<TElement> boundCollection;
+            CollectionResult result = null;
             if (valueProviderResult == null)
             {
-                boundCollection = await BindComplexCollection(bindingContext);
+                result = await BindComplexCollection(bindingContext);
+                boundCollection = result.Model;
             }
             else
             {
-                boundCollection = await BindSimpleCollection(
+                result = await BindSimpleCollection(
                     bindingContext,
                     valueProviderResult.RawValue,
                     valueProviderResult.Culture);
+                boundCollection = result.Model;
```

Comment:
ModelBindingResult has more than what is required(viz. properties like IsModelSet and the Key), moreover the model is loosely typed in MBR, resulting in an unnecessary cast.